File: packages/app/db/template.go
package db

import (
	"time"

	"gorm.io/gorm"
)

type Template struct {
	ID        string `gorm:"primarykey;not null"`
	CreatedAt time.Time
	UpdatedAt time.Time      `gorm:"index"`
	DeletedAt gorm.DeletedAt `gorm:"index"`

	ModelID string `gorm:"index"`
	Model   Model  `gorm:"constraint:OnDelete:CASCADE"`

	Name   string `gorm:"index"`
	Front  string
	Back   string
	Shared string
}

func (Template) Tidy(tx *gorm.DB) error {
	if r := tx.
		Where(`([name] IS NOT NULL AND ROWID NOT IN (SELECT ROWID FROM template WHERE [name] IS NOT NULL GROUP BY model_id, [name]))`).
		Or(`(front IS NULL AND model_id IN (SELECT id FROM model WHERE front IS NULL))`).
		Delete(&Template{}); r.Error != nil {
		return r.Error
	}

	return nil
}

File: packages/electron/src/db/template.ts
import {
  DateType,
  Entity,
  Index,
  ManyToOne,
  PrimaryKey,
  Property
} from '@mikro-orm/core'
import shortUUID from 'short-uuid'

import { Model } from './model'

@Entity()
export class Template {
  @PrimaryKey()
  id: string = shortUUID.generate()

  @Property({
    type: DateType
  })
  createdAt: Date = new Date()

  @Property({
    type: DateType,
    onUpdate: () => new Date()
  })
  updatedAt: Date = new Date()

  @ManyToOne(() => Model, { nullable: true })
  @Index()
  model?: Model

  @Property()
  @Index()
  name: string = ''

  @Property()
  front: string = ''

  @Property()
  back: string = ''

  @Property()
  shared: string = ''

  @Property()
  if: string = ''
}
